Generate the various TypeId enums in codegen

nox opened the pull request:
This frees us forever from caring about maintaining these enums. The last commit removes their use from the initialisation of interface objects derived from Node.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/servo/servo/7873)

<!-- Reviewable:end -->

nox committed:
Generate the TypeId enums in codegen

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify **unsafe code**. Please review it carefully!
* These commits modify layout code, but no reftests are modified. Please consider adding a reftest!

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7807) made this pull request unmergeable. Please resolve the merge conflicts.

michaelwu commented:
This is useful for me. I'd be happy with just the first commit, if that could be landed faster.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7875) made this pull request unmergeable. Please resolve the merge conflicts.

nox commented:
@Ms2ger 

```
AttrBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
BlobBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CSSBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CSSGroupingRuleBinding.rs:    type_id: None,
CSSImportRuleBinding.rs:    type_id: None,
CSSMarginRuleBinding.rs:    type_id: None,
CSSMediaRuleBinding.rs:    type_id: None,
CSSNamespaceRuleBinding.rs:    type_id: None,
CSSPageRuleBinding.rs:    type_id: None,
CSSRuleBinding.rs:    type_id: None,
CSSRuleListBinding.rs:    type_id: None,
CSSStyleDeclarationBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CSSStyleRuleBinding.rs:    type_id: None,
CSSStyleSheetBinding.rs:    type_id: None,
CanvasGradientBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CanvasPatternBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CanvasRenderingContext2DBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CharacterDataBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Abstract,
CloseEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::CloseEvent),
CommentBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::CharacterData(::dom::bindings::codegen::InheritTypes::CharacterDataTypeId::Comment))),
ConsoleBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CryptoBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
CustomEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::CustomEvent),
DOMExceptionBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMImplementationBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMParserBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMPointBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::DOMPointReadOnly(::dom::bindings::codegen::InheritTypes::DOMPointReadOnlyTypeId::DOMPoint),
DOMPointReadOnlyBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMRectBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMRectListBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMStringMapBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DOMTokenListBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
DedicatedWorkerGlobalScopeBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::WorkerGlobalScope(::dom::bindings::codegen::InheritTypes::WorkerGlobalScopeTypeId::DedicatedWorkerGlobalScope)),
DocumentBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Document)),
DocumentFragmentBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::DocumentFragment)),
DocumentTypeBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::DocumentType)),
ElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::Element))),
ErrorEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::ErrorEvent),
EventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
EventTargetBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
FileBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Blob(::dom::bindings::codegen::InheritTypes::BlobTypeId::File),
FileListBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
FileReaderBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::FileReader),
FormDataBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
HTMLAnchorElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLAnchorElement)))),
HTMLAppletElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLAppletElement)))),
HTMLAreaElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLAreaElement)))),
HTMLAudioElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLMediaElement(::dom::bindings::codegen::InheritTypes::HTMLMediaElementTypeId::HTMLAudioElement))))),
HTMLBRElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLBRElement)))),
HTMLBaseElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLBaseElement)))),
HTMLBodyElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLBodyElement)))),
HTMLButtonElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLButtonElement)))),
HTMLCanvasElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLCanvasElement)))),
HTMLCollectionBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
HTMLDListElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLDListElement)))),
HTMLDataElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLDataElement)))),
HTMLDataListElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLDataListElement)))),
HTMLDialogElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLDialogElement)))),
HTMLDirectoryElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLDirectoryElement)))),
HTMLDivElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLDivElement)))),
HTMLElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLElement)))),
HTMLEmbedElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLEmbedElement)))),
HTMLFieldSetElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLFieldSetElement)))),
HTMLFontElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLFontElement)))),
HTMLFormElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLFormElement)))),
HTMLFrameElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLFrameElement)))),
HTMLFrameSetElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLFrameSetElement)))),
HTMLHRElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLHRElement)))),
HTMLHeadElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLHeadElement)))),
HTMLHeadingElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLHeadingElement)))),
HTMLHtmlElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLHtmlElement)))),
HTMLIFrameElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLIFrameElement)))),
HTMLImageElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLImageElement)))),
HTMLInputElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLInputElement)))),
HTMLLIElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLLIElement)))),
HTMLLabelElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLLabelElement)))),
HTMLLegendElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLLegendElement)))),
HTMLLinkElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLLinkElement)))),
HTMLMapElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLMapElement)))),
HTMLMediaElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Abstract,
HTMLMetaElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLMetaElement)))),
HTMLMeterElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLMeterElement)))),
HTMLModElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLModElement)))),
HTMLOListElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLOListElement)))),
HTMLObjectElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLObjectElement)))),
HTMLOptGroupElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLOptGroupElement)))),
HTMLOptionElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLOptionElement)))),
HTMLOutputElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLOutputElement)))),
HTMLParagraphElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLParagraphElement)))),
HTMLParamElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLParamElement)))),
HTMLPreElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLPreElement)))),
HTMLProgressElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLProgressElement)))),
HTMLQuoteElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLQuoteElement)))),
HTMLScriptElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLScriptElement)))),
HTMLSelectElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLSelectElement)))),
HTMLSourceElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLSourceElement)))),
HTMLSpanElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLSpanElement)))),
HTMLStyleElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLStyleElement)))),
HTMLTableCaptionElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableCaptionElement)))),
HTMLTableCellElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Abstract,
HTMLTableColElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableColElement)))),
HTMLTableDataCellElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableCellElement(::dom::bindings::codegen::InheritTypes::HTMLTableCellElementTypeId::HTMLTableDataCellElement))))),
HTMLTableElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableElement)))),
HTMLTableHeaderCellElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableCellElement(::dom::bindings::codegen::InheritTypes::HTMLTableCellElementTypeId::HTMLTableHeaderCellElement))))),
HTMLTableRowElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableRowElement)))),
HTMLTableSectionElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTableSectionElement)))),
HTMLTemplateElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTemplateElement)))),
HTMLTextAreaElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTextAreaElement)))),
HTMLTimeElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTimeElement)))),
HTMLTitleElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTitleElement)))),
HTMLTrackElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLTrackElement)))),
HTMLUListElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLUListElement)))),
HTMLUnknownElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLUnknownElement)))),
HTMLVideoElementBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::Element(::dom::bindings::codegen::InheritTypes::ElementTypeId::HTMLElement(::dom::bindings::codegen::InheritTypes::HTMLElementTypeId::HTMLMediaElement(::dom::bindings::codegen::InheritTypes::HTMLMediaElementTypeId::HTMLVideoElement))))),
ImageDataBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
KeyboardEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::UIEvent(::dom::bindings::codegen::InheritTypes::UIEventTypeId::KeyboardEvent)),
LocationBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
MessageEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::MessageEvent),
MouseEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::UIEvent(::dom::bindings::codegen::InheritTypes::UIEventTypeId::MouseEvent)),
NamedNodeMapBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
NavigatorBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
NodeBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Abstract,
NodeIteratorBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
NodeListBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
PerformanceBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
PerformanceTimingBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
ProcessingInstructionBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::CharacterData(::dom::bindings::codegen::InheritTypes::CharacterDataTypeId::ProcessingInstruction))),
ProgressEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::ProgressEvent),
RangeBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
ScreenBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
ServoHTMLParserBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
StorageBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
StorageEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::StorageEvent),
StyleSheetBinding.rs:    type_id: None,
StyleSheetListBinding.rs:    type_id: None,
TestBindingBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
TestBindingProxyBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::TestBinding(::dom::bindings::codegen::InheritTypes::TestBindingTypeId::TestBindingProxy),
TextBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Node(::dom::bindings::codegen::InheritTypes::NodeTypeId::CharacterData(::dom::bindings::codegen::InheritTypes::CharacterDataTypeId::Text))),
TextDecoderBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
TextEncoderBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
TreeWalkerBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
UIEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::UIEvent(::dom::bindings::codegen::InheritTypes::UIEventTypeId::UIEvent)),
URLBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
URLSearchParamsBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
ValidityStateBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WebGLActiveInfoBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WebGLBufferBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::WebGLObject(::dom::bindings::codegen::InheritTypes::WebGLObjectTypeId::WebGLBuffer),
WebGLContextEventBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Event(::dom::bindings::codegen::InheritTypes::EventTypeId::WebGLContextEvent),
WebGLFramebufferBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::WebGLObject(::dom::bindings::codegen::InheritTypes::WebGLObjectTypeId::WebGLFramebuffer),
WebGLObjectBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WebGLProgramBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::WebGLObject(::dom::bindings::codegen::InheritTypes::WebGLObjectTypeId::WebGLProgram),
WebGLRenderbufferBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::WebGLObject(::dom::bindings::codegen::InheritTypes::WebGLObjectTypeId::WebGLRenderbuffer),
WebGLRenderingContextBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WebGLShaderBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::WebGLObject(::dom::bindings::codegen::InheritTypes::WebGLObjectTypeId::WebGLShader),
WebGLShaderPrecisionFormatBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WebGLTextureBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::WebGLObject(::dom::bindings::codegen::InheritTypes::WebGLObjectTypeId::WebGLTexture),
WebGLUniformLocationBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WebSocketBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::WebSocket),
WindowBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Window),
WorkerBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::Worker),
WorkerGlobalScopeBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Abstract,
WorkerLocationBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
WorkerNavigatorBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Alone,
XMLHttpRequestBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::XMLHttpRequestEventTarget(::dom::bindings::codegen::InheritTypes::XMLHttpRequestEventTargetTypeId::XMLHttpRequest)),
XMLHttpRequestEventTargetBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::Abstract,
XMLHttpRequestUploadBinding.rs:    type_id: ::dom::bindings::codegen::InheritTypes::TopTypeId::EventTarget(::dom::bindings::codegen::InheritTypes::EventTargetTypeId::XMLHttpRequestEventTarget(::dom::bindings::codegen::InheritTypes::XMLHttpRequestEventTargetTypeId::XMLHttpRequestUpload)),
```

nox commented:
https://gist.github.com/nox/5778ada41c9da182d2fe

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7937) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
Review status: 0 of 106 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 1735 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0BnIj03RdT0_WWKD6o)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L1735)):</sup>
`name` is only used inside the if, so move it there.
<!-- Reviewable comment -K0BnIj1dAB6KDZRxsMs:-2139970417 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 1740 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0BtXxLtxMbq4GF33Eu)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L1740)):</sup>
Maybe this would be better as
```python
reversed_protochain = list(reversed(protochain))
for (prev_proto, proto) in zip(reversed_protochain, reversed_protochain[1:]):
    ...
```

Maybe.
<!-- Reviewable comment -K0BtXxMB_5n8nqhoZ6t:-1994053323 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 1741 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0BuDaT1TlG7sJ6nxVg)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L1741)):</sup>
Also, maybe "parent" and "child" would be better than "proto" and "prev_p.roto"
<!-- Reviewable comment -K0BuDaUuRylhGAQu2UG:1860207879 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5873 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0BzxRbpxbAF9TnS5Bp)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L5873)):</sup>
Remove
<!-- Reviewable comment -K0BzxRbpxbAF9TnS5Bq:2097897961 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5882 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0BzybzJaOf_Oru-SsK)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L5882)):</sup>
This adds descriptors where `len(descriptor.prototypeChain) == 1`, which should be `Alone`, I think. Is the correct condition `upcast and not downcast`?
<!-- Reviewable comment -K0BzybzJaOf_Oru-SsL:44613850 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5977 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0Bproswzs6QGwFx7FG)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L5977)):</sup>
Manual indentation kinda sucks here.
<!-- Reviewable comment -K0Bprotd9eEiFsVHzLV:900999080 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5978 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0BpuH4kxAFvmZAB0ES)** ([raw file](https://github.com/servo/servo/blob/64afde2de286dae3fb1c1d6fc54765d7dbe6daae/components/script/dom/bindings/codegen/CodegenRust.py#L5978)):</sup>
I'd also like doc comments for all those.
<!-- Reviewable comment -K0BpuH4kxAFvmZAB0ET:2018557164 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

nox commented:
@bors-servo p=1

So that it takes priority when it gets reviewed.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7953) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
I think I'm okay with the first commit with these changes.
<!-- Reviewable comment -K0WsnhcDi-Rm98ZVurx:224398081 -->

---

Review status: 0 of 106 files reviewed at latest revision, 9 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5882 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WoMpuq-gsQBMncQUc)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5882)):</sup>
I know that this refers to the parent interface, but it deserves a comment. Maybe it's more readable as `descriptor.interface.parent.identifier.name`? Maybe we could have a method on `Descriptor` along those lines?
<!-- Reviewable comment -K0WoMpuq-gsQBMncQUd:1410542089 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5968 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WoxDybBdeCRGriAjd)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5968)):</sup>
Not entirely sure what's going on in this function.
<!-- Reviewable comment -K0WoxDybBdeCRGriAje:-2087706135 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5971 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WouyL4DT7aWAQHdcX)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5971)):</sup>
No else-after-return.
<!-- Reviewable comment -K0WouyMlAba84PDTjBw:2027570483 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5974 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0Wq3Wuzyar_s5T85N3)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5974)):</sup>
What does `types` mean here?
<!-- Reviewable comment -K0Wq3Wuzyar_s5T85N4:1755251677 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5976 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0Wp1T9FcebRAEhG2mJ)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5976)):</sup>
Maybe `variants` should be a list of `(documentation, variant)` tuples; I think that'd read better.
<!-- Reviewable comment -K0Wp1T9FcebRAEhG2mK:23639352 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5988 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WpeyXGkuPb_4M3jIY)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5988)):</sup>
Eek, shadowing
<!-- Reviewable comment -K0WpeyYTAxqoqXEu1EX:-1650239608 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5991 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WqApIoCJ-uHVikmrH)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5991)):</sup>
Move this formatting down too.
<!-- Reviewable comment -K0WqApIoCJ-uHVikmrI:-347614667 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 6000 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WsbcSEOUyCbSTgDaI)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L6000)):</sup>
I think these can be inherent implementations.
<!-- Reviewable comment -K0WsbcTULL2Qa3R9Ukg:-259733492 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 6002 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0Wr73A3KVbKDcsdlkZ)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L6002)):</sup>
`-> &'static`?
<!-- Reviewable comment -K0Wr73Bn4SjGX7oDNs2:45627458 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

nox commented:
Review status: 0 of 107 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5968 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0WoxDybBdeCRGriAjd)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5968)):</sup>
Added a comment.
<!-- Reviewable comment -K0WvUl_vBMjMiiePy-w:-921755437 -->

---

<sup>**[components/script/dom/bindings/codegen/CodegenRust.py, line 5974 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0Wq3Wuzyar_s5T85N3)** ([raw file](https://github.com/servo/servo/blob/91db0221cd440822c914eed74179eb11a46d684e/components/script/dom/bindings/codegen/CodegenRust.py#L5974)):</sup>
Renamed the variable to `typeIdCode`.
<!-- Reviewable comment -K0X4fGG6U2BlqlBTeaf:1995750081 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

nox commented:
@Ms2ger I removed the two latter commits from the branch and addressed your comments.

bors-servo commented:
:umbrella: The latest upstream changes (presumably #7829) made this pull request unmergeable. Please resolve the merge conflicts.

Ms2ger commented:
-S-awaiting-review +S-needs-code-changes 
<!-- Reviewable comment -K0b-PeJIsdh44ew9jdW:1458266033 -->

---

Reviewed 7 of 106 files at r1, 4 of 30 files at r4, 96 of 96 files at r5.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/bindings/utils.rs, line 159 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0ayTj55MdFjbbMvOLa)** ([raw file](https://github.com/servo/servo/blob/56ba0d39ee6e4920a32383bcd07b1fee586cf2f6/components/script/dom/bindings/utils.rs#L159)):</sup>
Nit: no need for "if this is derived from any other interface" anymore
<!-- Reviewable comment -K0ayTj6ETPrqZ1flcLk:-742345360 -->

---

<sup>**[components/script/dom/eventtarget.rs, line 29 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0az-CKkVqxZbXohAo9)** ([raw file](https://github.com/servo/servo/blob/56ba0d39ee6e4920a32383bcd07b1fee586cf2f6/components/script/dom/eventtarget.rs#L29)):</sup>
Please revert those changes.
<!-- Reviewable comment -K0az-CKkVqxZbXohAoA:1400417833 -->

---

<sup>**[components/script/dom/file.rs, line 20 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0azDgiqIKJVpG6II75)** ([raw file](https://github.com/servo/servo/blob/56ba0d39ee6e4920a32383bcd07b1fee586cf2f6/components/script/dom/file.rs#L20)):</sup>
I would prefer keeping this where it was for now.
<!-- Reviewable comment -K0azDgiqIKJVpG6II76:605142028 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

nox commented:
Done, LGTM!
<!-- Reviewable comment -K0b1hEMd3ghFYkY3e6r:895270002 -->

---

Review status: 103 of 107 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

<sup>**[components/script/dom/bindings/utils.rs, line 159 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0ayTj55MdFjbbMvOLa)** ([raw file](https://github.com/servo/servo/blob/56ba0d39ee6e4920a32383bcd07b1fee586cf2f6/components/script/dom/bindings/utils.rs#L159)):</sup>
Good catch.
<!-- Reviewable comment -K0b0EY4as0ycQt-8DHC:422879057 -->

---

<sup>**[components/script/dom/eventtarget.rs, line 29 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0az-CKkVqxZbXohAo9)** ([raw file](https://github.com/servo/servo/blob/56ba0d39ee6e4920a32383bcd07b1fee586cf2f6/components/script/dom/eventtarget.rs#L29)):</sup>
Good catch.
<!-- Reviewable comment -K0b0HYaZ76aMXy9z1rX:422879057 -->

---

<sup>**[components/script/dom/file.rs, line 20 \[r5\]](https://reviewable.io:443/reviews/servo/servo/7873#-K0azDgiqIKJVpG6II75)** ([raw file](https://github.com/servo/servo/blob/56ba0d39ee6e4920a32383bcd07b1fee586cf2f6/components/script/dom/file.rs#L20)):</sup>
Done.
<!-- Reviewable comment -K0b1ePOTGDtqWxVui85:-499146009 -->

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

Ms2ger commented:
-S-awaiting-review +S-awaiting-merge 
@bors-servo r+
Thank you!
<!-- Reviewable comment -K0b2be0pMArIKoep-3V:-62040282 -->

---

Reviewed 4 of 4 files at r6.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit bea04e7 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger bea04e7825f52d08d25b27a1839411dc34059b95 -->

bors-servo commented:
:hourglass: Testing commit bea04e7 with merge 467d3d2...

bors-servo commented:
:broken_heart: Test failed - [linux-rel](http://build.servo.org/builders/linux-rel/builds/518)

Ms2ger commented:
@bors-servo r+
-S-awaiting-review +S-awaiting-merge 
<!-- Reviewable comment -K0bFDyWbWLoC3mqxx9-:-819973780 -->

---

Reviewed 2 of 2 files at r7.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks pending.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7873)
<!-- Sent from Reviewable.io -->

bors-servo commented:
:pushpin: Commit aab2c40 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger aab2c40389c71c4b0db3ffa75bdec99cd440b8af -->

bors-servo commented:
:hourglass: Testing commit aab2c40 with merge 80be62a...

bors-servo commented:
:broken_heart: Test failed - [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/555)

nox commented:
@bors-servo: r+ #7787

bors-servo commented:
:pushpin: Commit aab2c40 has been approved by `nox`

<!-- @bors-servo r=nox aab2c40389c71c4b0db3ffa75bdec99cd440b8af -->

nox commented:
Err.

nox commented:
@bors-servo r-

nox commented:
@bors-servo r=Ms2ger

bors-servo commented:
:pushpin: Commit aab2c40 has been approved by `Ms2ger`

<!-- @bors-servo r=Ms2ger aab2c40389c71c4b0db3ffa75bdec99cd440b8af -->

nox commented:
@bors-servo retry

bors-servo commented:
:hourglass: Testing commit aab2c40 with merge 32daa17...

bors-servo commented:
:sunny: Test successful - [android](http://build.servo.org/builders/android/builds/3426), [gonk](http://build.servo.org/builders/gonk/builds/2586), [linux-dev](http://build.servo.org/builders/linux-dev/builds/504), [linux-rel](http://build.servo.org/builders/linux-rel/builds/521), [mac-dev-ref-unit](http://build.servo.org/builders/mac-dev-ref-unit/builds/539), [mac-rel-css](http://build.servo.org/builders/mac-rel-css/builds/521), [mac-rel-wpt](http://build.servo.org/builders/mac-rel-wpt/builds/556)

frewsxcv commented:
![](https://media.giphy.com/media/YYD3fLEOdcOv6/giphy.gif)
